website/blog/2024/12-13-scrape-google-maps-using-python/index.md

```diff
@@ -0,0 +1,481 @@
+
+![Hotel Review Count](./img/scrape-google-maps-with-crawlee-screenshot-reviews.webp)
+
+**Hotel URL:**
+
+![Hotel URL](./img/scrape-google-maps-with-crawlee-screenshot-url.webp)
+
+**Hotel Price:**
+
+![Hotel Price](./img/scrape-google-maps-with-crawlee-screenshot-price.webp)
+
+**Hotel amenities:**
+
+This returns multiple elements as each hotel has several amenities. We'll need to iterate through these.
+
+![Hotel amenities](./img/scrape-google-maps-with-crawlee-screenshot-amenities.webp)
+
+**Quick tips:**
+
+- Always verify these selectors before scraping, as Google might update them.
+- Use Chrome DevTools (F12) to inspect elements and confirm selectors.
+- Some elements might not be present for all hotels (like prices during the off-season).
+
+### 4. Scraping Google Maps data using identified selectors
+
+Let's build a scraper to extract detailed hotel information from Google Maps. First, create the core scraping function to handle data extraction.
+
+*gmap_scraper.py:*
+
+```python
+async def scrape_google_maps(context) -> None:
+    page = context.page
+    print(f"\nProcessing URL: {context.request.url}\n")
+
+    # Wait for content to load
+    await page.wait_for_selector(".Nv2PK", timeout=30000)
+    await page.wait_for_timeout(2000)
+
+    # Get all hotel listings
+    listings = await page.query_selector_all(".Nv2PK")
+    print(f"Found {len(listings)} hotels\n")
+
+    # Process each hotel listing
+    for listing in listings:
+        # Extract details for each listing
+        data = {
+            "name": (
+                await (await listing.query_selector(".qBF1Pd")).inner_text()
+                if await listing.query_selector(".qBF1Pd")
+                else None
+            ),
+            "rating": (
+                await (await listing.query_selector(".MW4etd")).inner_text()
+                if await listing.query_selector(".MW4etd")
+                else "N/A"
+            ),
+            "reviews": (
+                (await (await listing.query_selector(".UY7F9")).inner_text())
+                .replace("(", "")
+                .replace(")", "")
+                if await listing.query_selector(".UY7F9")
+                else "N/A"
+            ),
+            "price": (
+                await (await listing.query_selector(".wcldff")).inner_text()
+                if await listing.query_selector(".wcldff")
+                else "N/A"
+            ),
+            "link": (
+                await (await listing.query_selector("a.hfpxzc")).get_attribute("href")
+                if await listing.query_selector("a.hfpxzc")
+                else "N/A"
+            ),
+            "amenities": [
+                await amenity.get_attribute("aria-label")
+                for amenity in await listing.query_selector_all(".dc6iWb")
+                if await amenity.get_attribute("aria-label")
+            ],
+        }
+
+        # Pretty-print the data
+        print(json.dumps(data, indent=4))
+        print("\n")
```

Comment:
This is not how it's supposed to be done - it'd be better to use `context.push_data(data)`